### ISSUE TITLE
Unable to train simple flair text classifier on 20 newsgroups dataset. Naive Bayes classifier outperforms

### ISSUE THREAD

**Issue by Solid7** [zalandoresearch/flair#1230]
First of all thanks for an excellent library, still i'm having some issues =).

I'm trying to train a simple text classifier on the '20 newsgroup dataset'. I'm not able to get F1 score higher than ~0.4. Naive Bayes classifier on same dataset reaches F1 score: ~0.8.

Simple naive bayes:

    from sklearn.datasets import fetch_20newsgroups
    from sklearn.naive_bayes import MultinomialNB
    from sklearn import metrics
    from sklearn.feature_extraction.text import TfidfVectorizer
    vectorizer = TfidfVectorizer()

    newsgroups_train = fetch_20newsgroups(subset='train')
    newsgroups_test = fetch_20newsgroups(subset='test')
    vectors_train = vectorizer.fit_transform(newsgroups_train.data)
    vectors_test = vectorizer.transform(newsgroups_test.data)
    
    clf = MultinomialNB(alpha=.01)
    clf.fit(vectors_train, newsgroups_train.target)
    pred = clf.predict(vectors_test)
    print('f1 score: ', metrics.f1_score(newsgroups_test.target, pred, average='macro'))
    
    import numpy as np
    def show_top10(classifier, vectorizer, categories):
        feature_names = np.asarray(vectorizer.get_feature_names())
        for i, category in enumerate(categories):
            top10 = np.argsort(classifier.coef_[i])[-10:]
            print("%s: %s" % (category, " ".join(feature_names[top10])))
    
    show_top10(clf, vectorizer, newsgroups_train.target_names)

Result

    f1 score:  0.8290659644474043

The Flair text classifier on same dataset:

    import pandas as pd
    import numpy as np
    from sklearn.datasets import fetch_20newsgroups
    from flair.data_fetcher import NLPTaskDataFetcher
    from flair.embeddings import (WordEmbeddings, 
    FlairEmbeddings, DocumentLSTMEmbeddings, RoBERTaEmbeddings, 
    BertEmbeddings, DocumentPoolEmbeddings)
    from flair.models import TextClassifier
    from flair.trainers import ModelTrainer

    # Avoid gpu bug
    import flair, torch
    flair.device = torch.device('cpu') 
    
    def create_dataset():
        newsgroups_train = fetch_20newsgroups(subset='train')
        newsgroups_test = fetch_20newsgroups(subset='test')
        labels = np.array(newsgroups_test.target_names)
    
        df_train = pd.DataFrame(dict(label=labels[newsgroups_train.target], text=newsgroups_train.data))
        df_train['label'] = '__label__' + df_train['label'].astype('str')
    
        df_test = pd.DataFrame(dict(label=labels[newsgroups_test.target], text=newsgroups_test.data))
        df_test['label'] = '__label__' + df_test['label'].astype('str')
    
        df_test = df_test.iloc[0:int(len(df_test)*0.5)]
        df_dev = df_test.iloc[int(len(df_test)*0.5):]
        
        for df, name in zip([df_train, df_test, df_dev], ['train', 'test', 'dev']):
            df.to_csv(name + '.csv', sep=' ', index=False, header=False, quotechar=' ')
    
        corpus = NLPTaskDataFetcher.load_classification_corpus('./', train_file='train.csv', 
        test_file='test.csv', dev_file='dev.csv')
    
        return corpus
    
    def train_flair():
        corpus = create_dataset()
    
        word_embeddings = [WordEmbeddings('glove'), FlairEmbeddings('news-forward-fast'), FlairEmbeddings('news-backward-fast')]
      
        document_embeddings = DocumentLSTMEmbeddings(word_embeddings, hidden_size=128, reproject_words=True)
    
        dictionary = corpus.make_label_dictionary()
        classifier = TextClassifier(document_embeddings, label_dictionary=dictionary, multi_label=False)
    
        trainer = ModelTrainer(classifier, corpus, use_tensorboard=True)
        trainer.train('resources/taggers/classy', 
            max_epochs=25, patience=1, learning_rate=0.2, 
            mini_batch_size=32, anneal_factor=0.5, embeddings_storage_mode='gpu')
    
    create_dataset()
    train_flair()
    

Output (after 58 epcohs):

    MICRO_AVG: acc 0.2429 - f1-score 0.3909
    MACRO_AVG: acc 0.2586 - f1-score 0.39669

I'm not sure what is going on here. If there is something wrong with my creation of the dataset for the Flair text classifier or the training/model code. I've never used Flair before. If anyone with experience with Flair could guide me in the right direction that would be super good.

**Comment by robertoptonova** [comment]
As per answer from @alanakbik : "Hi @Solid7 it looks like the files you are generating are not proper FastText formatted files. In FastText format, each line is one data point, i.e. you can have no line breaks and each line needs to start with a label formatted label."

https://github.com/zalandoresearch/flair/issues/1229#issuecomment-544995420

Think this is my issue.

**Comment by Solid7** [comment]
As per answer from @alanakbik : "Hi @Solid7 it looks like the files you are generating are not proper FastText formatted files. In FastText format, each line is one data point, i.e. you can have no line breaks and each line needs to start with a label formatted label."

https://github.com/zalandoresearch/flair/issues/1229#issuecomment-544995420

**Comment by DecentMakeover** [comment]
@Solid7 did you manage to run this with good score?if so can you post code to reproduce, i am getting 66 on newsgroup.

Thanks.

**Comment by Solid7** [comment]
I don't have the code but I think fixing the line endings made it much better. Should be fairly simple, think I just removed all line endings.

**Comment by DecentMakeover** [comment]
You mean new lines ?

On 19-Dec-2019, at 1:17 AM, Solid7 <notifications@github.com<mailto:notifications@github.com>> wrote:


I don't have the code but I think fixing the line endings made it much better. Should be fairly simple, think I just removed all line endings.
